default.nix

```diff
@@ -0,0 +1,49 @@
+# Default import pinned pkgs
+{ pkgsSrc ? (import ./nix/pkgs.nix {}).pkgsSrc
+, pkgs ? (import ./nix/pkgs.nix { inherit pkgsSrc; }).pkgs
+, doCheck ? true
+}: with pkgs;
+
+let
+  # Get contract dependencies from lock file
+  inherit (callPackage ./nix/dapp.nix {}) specs packageSpecs;
+  inherit (specs.this) deps;
+
+  # Import deploy scripts from dss-deploy
+  dss-deploy = import deps.dss-deploy.src' {};
+
+  # Create derivations from lock file data
+  packages = packageSpecs (deps // {
+    # Set specific solc versions for some contract derivations
+    multicall = deps.multicall // { solc = solc-versions.solc_0_4_25; };
+    vote-proxy = deps.vote-proxy // { solc = solc-versions.solc_0_4_25; };
+  });
+in makerScriptPackage {
+  name = "testchain-dss-deploy-scripts";
+
+  # Specify files to add to build environment
+  src = lib.sourceByRegex ./. [
+    ".*deploy"
+    ".*\.json"
+    ".*scripts.*"
+    ".*lib.*"
+  ];
+
+  solidityPackages = builtins.attrValues packages;
+
+  extraBins = [
+    dss-deploy
+  ];
+
+  # Patch scripts by removing `cd` commands and adding DAPP_OUT env so that
+  # `dapp create` can find the right binaries even though no submodules have
+  # been downloaded inside the Nix build environment.
+  patchBin = writeScript "remove-cd" ''
+    #!${stdenv.shell}
+    exec ${perl}/bin/perl -pe '
+      s|^(\s*)cd\s+(?:.*/)*([^/\n\r\s;&\|\.]+)|\1export DAPP_OUT=\$DAPP_LIB/\2/out;|;
```

Comment:
Wow.

Reply:
Some great software engineering right there :P

Reply:
I think we should drop support for bare running of the scripts and just remove the `cd`

Reply:
@asymmetric why? It think it is the easiest way for developing and testing. I honestly prefer to keep supporting the direct execution.

Reply:
Because supporting both has a cost (see e.g. that regexp - updating/debugging it could be a huge time sink!), and I think eventually we have to choose what tool we want to use and just go for it.

Reply:
I'm not saying we should do it now, let's try this in the wild for a few iterations, but then let's make a decision.

Reply:
Should definitely evaluate this first and make a document collecting pros/cons and future improvements.

Reply:
I think we can properly solve it with environment variables as @icetan did for `dss-deploy` in the next release. If we encapsulate those `cd` with `dapp build` in those vars we could avoid extra complexity.

Reply:
@gbalabasquer good idea, could add an env variable (e.g. `BUILD_CMD`) that defaults to:

```
BUILD_CMD=${BUILD_CMD:-'
  dappBuildCmd() {
    cd "$CONTRACT_DIR/$1"
    dapp $2 build
  }
'}
```
And is then added to `lib/common.sh` then used like: `dappBuildCmd "line-spell"`.

The Nix derivation can then inject:

```
BUILD_CMD='
  dappBuildCmd() {
    export DAPP_OUT="$DAPP_LIB/dapp/$1/out"
  }
'
```

No regex needed.

Reply:
How is the string then evaluated and turned into a function? `eval`?

Reply:
> How is the string then evaluated and turned into a function? `eval`?

Yes `lib/common.sh` would have a line with `eval "$BUILD_CMD"`.

Reply:
Here is a better solution though: https://github.com/makerdao/testchain-dss-deployment-scripts/pull/34